File: regions/system/library/typescript/app-updater/src/checksum.ts
import { createHash } from 'node:crypto';
import { createReadStream } from 'node:fs';
import { ChecksumError } from './error.js';

/** ファイルの SHA-256 チェックサムを計算・検証するユーティリティ */
export class ChecksumVerifier {
  /** ファイルの SHA-256 チェックサムを計算して16進数文字列で返す */
  static async calculate(filePath: string): Promise<string> {
    return new Promise((resolve, reject) => {
      const hash = createHash('sha256');
      const stream = createReadStream(filePath);
      stream.on('data', (chunk) => hash.update(chunk));
      stream.on('end', () => resolve(hash.digest('hex')));
      stream.on('error', reject);
    });
  }

  /** ファイルのチェックサムが期待値と一致するかを検証する */
  static async verify(filePath: string, expected: string): Promise<boolean> {
    const actual = await ChecksumVerifier.calculate(filePath);
    return actual === expected.toLowerCase();
  }

  /** ファイルのチェックサムが期待値と一致しない場合は例外をスローする */
  static async verifyOrThrow(filePath: string, expected: string): Promise<void> {
    const verified = await ChecksumVerifier.verify(filePath, expected);
    if (!verified) {
      throw new ChecksumError('File checksum did not match.');
    }
  }
}

File: regions/system/library/rust/app-updater/src/checksum.rs
use sha2::{Digest, Sha256};
use std::path::Path;
use tokio::fs;

use crate::error::AppUpdaterError;

/// ファイルの SHA-256 チェックサムを計算・検証するユーティリティ
pub struct ChecksumVerifier;

impl ChecksumVerifier {
    /// ファイルの SHA-256 チェックサムを計算して16進数文字列で返す
    pub async fn calculate(file_path: &Path) -> Result<String, AppUpdaterError> {
        let bytes = fs::read(file_path).await?;
        let mut hasher = Sha256::new();
        hasher.update(&bytes);
        let result = hasher.finalize();
        Ok(format!("{result:x}"))
    }

    /// ファイルのチェックサムが期待値と一致するかを検証する
    pub async fn verify(file_path: &Path, expected: &str) -> Result<bool, AppUpdaterError> {
        let actual = Self::calculate(file_path).await?;
        Ok(actual == expected.to_lowercase())
    }

    /// ファイルのチェックサムが期待値と一致しない場合はエラーを返す
    pub async fn verify_or_error(file_path: &Path, expected: &str) -> Result<(), AppUpdaterError> {
        let verified = Self::verify(file_path, expected).await?;
        if !verified {
            return Err(AppUpdaterError::Checksum(
                "file checksum did not match expected value".to_string(),
            ));
        }
        Ok(())
    }
}
